plugins/outputs/mqtt/topic_name_generator.go

```diff
@@ -0,0 +1,49 @@
+package mqtt
+
+import (
+	"strings"
+	"text/template"
+
+	"github.com/influxdata/telegraf"
+)
+
+type TopicNameGenerator struct {
+	Hostname    string
+	TopicPrefix string
+	metric      telegraf.Metric
+	template    *template.Template
+}
+
+func NewTopicNameGenerator(hostname string, topicPrefix string, temp *template.Template) *TopicNameGenerator {
+	return &TopicNameGenerator{Hostname: hostname, TopicPrefix: topicPrefix, template: temp}
+}
+
+func (t *TopicNameGenerator) Tag(key string) string {
+	tagString, _ := t.metric.GetTag(key)
+	return tagString
+}
+
+func (t *TopicNameGenerator) PluginName() string {
+	return t.metric.Name()
+}
+
+func (t *TopicNameGenerator) Generate(m telegraf.Metric) string {
+	t.metric = m
```

Comment:
@srebhan I think I have done something close to what you originally requested.
The difficulty here is that I can't make the metric variable a transient one. It has to be part of the struct in order to make the PluginName and the Tag able to work.

Reply:
Putting this into the metric is OK I think.